### PR TITLE
Fix null vector normalization in ship select screen

### PR DESCRIPTION
**Pull request by asarium**
`vm_vector_2_matrix` was improperly guarded against bad parameters. In the
case I encountered the `uvec` was the negative of `fvec` which produced a
null vector when the cross product of the two vector was constructed.

These changes should fix that bug by ignoring the `uvec` parameter. Since
`vm_vector_2_matrix` and `vm_vector_2_matrix_norm` are basically the same, I
converted the former into an alias of the latter and only fixed that
code.

### CODE REVIEW

File: code/math/vecmat.cpp [hunk 1]
```diff
@@ -813,60 +813,45 @@ void vm_vector_2_matrix_gen_vectors(matrix *m)
...
 
 	*zvec = *fvec;
 
 	if (uvec == NULL) {
 		if (rvec == NULL) {     //just forward vec
-			vm_vector_2_matrix_gen_vectors(m);
+			vm_vector_2_matrix_gen_vectors(&temp);
 		}
 		else {                      //use right vec
+			*xvec = *rvec;
```
Comment (asarium) [inline code/math/vecmat.cpp]: I'm not entirely sure about this change but as far as I can tell the `rvec` value was never used in the previous version.

File: code/math/vecmat.cpp [hunk 2]
```diff
@@ -877,14 +862,29 @@ matrix *vm_vector_2_matrix_norm(matrix *m, const vec3d *fvec, const vec3d *uvec,
 		}
 	}
 	else {      //use up vec
+		*yvec = *uvec;
```
Comment (asarium) [inline code/math/vecmat.cpp]: Same as above.

File: code/math/vecmat.cpp (hunk 1 above)
Comment (DahBlount) [inline code/math/vecmat.cpp]: I don't believe it was. Regardless, if you know the handedness of the coordinate system and you have two orthogonal vectors and know which one is the fvec, rvec, and/or uvec, you can generate the matrix.

**Comment by MageKing17** [comment]
> In the case I encountered the `uvec` was the negative of `fvec` which produced a null vector when the cross product of the two vector was constructed.

Shouldn't something be done about the code that calls it with such vectors?

**Comment by asarium** [comment]
The code which caused the issue in my case was the shadow setup code where the up vector of the eye matrix happend to be the same as the light vector. I don't think there is a universal way of fixing such code.

**Comment by DahBlount** [comment]
The change to vm_vector_2_matrix should be sufficient imo. There's not much we can do about vectors that incidentally happen to be the equal and/or opposite when constructing a matrix except ignore the one that doesn't matter as much. Regardless, this does afford some extra protection within the functions themselves to avoid null vector issues overall.

**Comment by z64555** [comment]
Ok, original code seems to have made the assumption that the input matrix was an Identity matrix, which explains why in vector_2_matrix_norm it was able to compute the matrix while outright ignoring the value of uvec (it just checked if it was nullptr or not).

New changes now respect the uvec, and should solve quite a few issues, and the added safety is quite welcome.

Before we should merge this in, however, the vector_2_matrix functions should have a serious doxy on them. The comments there leave out a lot of important info (like the fact that it assumed the matrix was an identity, and thus was an in/out param)

**Comment by asarium** [comment]
Adding documentation is not in the scope of this changeset since these changes only fix the null vector normalization. If you want to add documentation then that should be done in a separate PR.

**Comment by z64555** [comment]
Very well, will make a new PR with doxy changes shortly.

**Pull request merged**